enterprise-suite/console_test/prometheus_test.go

```diff
@@ -0,0 +1,264 @@
+		appInstancesQuery := fmt.Sprintf("count( count by (instance) (ohai{es_workload=\"es-test\", namespace=\"%v\"}) ) == 2", promTestNamespace)
+		err := util.WaitUntilTrue(func() bool {
+			return prom.HasData(appInstancesQuery)
+		})
+		Expect(err).To(Succeed())
+
+		// Pod discovery - automatic metrics should gain an es_monitor_type label when a custom monitor is created
+		Expect(esMonitor.Make("es-test/my_custom_monitor", "up")).To(Succeed())
+		err = util.WaitUntilTrue(func() bool {
+			return prom.HasModel("my_custom_monitor")
+		})
+		Expect(err).To(Succeed())
+
+		// App via 'Pod' service discovery with multiple metric ports
+		appInstancesWithMultiplePortsQuery := fmt.Sprintf("count( count by (instance) (ohai{es_workload=\"es-test-with-multiple-ports\", namespace=\"%v\"}) ) == 4", promTestNamespace)
+		err = util.WaitUntilTrue(func() bool {
+			return prom.HasData(appInstancesWithMultiplePortsQuery)
+		})
+		Expect(err).To(Succeed())
+
+		// App via 'Service' service discovery
+		appInstancesViaServiceQuery := fmt.Sprintf("count( count by (instance) (ohai{es_workload=\"es-test-via-service\", namespace=\"%v\"}) ) == 2", promTestNamespace)
+		err = util.WaitUntilTrue(func() bool {
+			return prom.HasData(appInstancesViaServiceQuery)
+		})
+		Expect(err).To(Succeed())
+
+		// 'Service' discovery - automatic metrics should gain an es_monitor_type label when a custom monitor is created
+		Expect(esMonitor.Make("es-test-via-service/my_custom_monitor_for_service", "up")).To(Succeed())
+		err = util.WaitUntilTrue(func() bool {
+			return prom.HasModel("my_custom_monitor_for_service")
+		})
+		Expect(err).To(Succeed())
+		Expect(prom.HasData("up{es_workload=\"es-test-via-service\", es_monitor_type=\"es-test-via-service\"}")).To(BeTrue())
+
+		// 'Service' discovery with only endpoints
+		err = util.WaitUntilTrue(func() bool {
+			return prom.HasData(fmt.Sprintf("count( count by (instance) (up{ "+
+				"job=\"kubernetes-services\", kubernetes_name=\"es-test-service-with-only-endpoints\", namespace=\"%v\""+
+				"}) ) == 1", promTestNamespace))
+		})
+		Expect(err).To(Succeed())
+
+		// kubernetes-cadvisor metrics should have an es_monitor_type label.
+		Expect(esMonitor.Make("es-test/es-monitor-type-test", "container_cpu_load_average_10s")).To(Succeed())
+		err = util.WaitUntilTrue(func() bool {
+			return prom.HasModel("es-monitor-type-test")
+		})
+		Expect(err).To(Succeed())
+
+		// Specific test for regression of es-backend/issues/430.
+		Expect(prom.HasData("{job=\"kubernetes-cadvisor\",es_monitor_type=\"es-test\"}")).To(BeTrue())
+
+		// Succeeds if all data for the workload es-test  has a matching es_monitor_type
+		// Note we're currently ignoring health metrics because 'bad' data can stick around for 15m given their time window.
+		err = util.WaitUntilTrue(func() bool {
+			return prom.HasData("{es_workload=\"es-test\", es_monitor_type!=\"es-test\", __name__!=\"health\"}")
+		})
+		Expect(err).To(Succeed())
+	})
```

Comment:
It would be a lot nicer if this was broken up into a few asserts per "It" block. It's much harder to see what we're testing with a giant block like this. It also makes the test failures/messages less obvious, since "it works" "failed" isn't very specific. I know it's just copied from what we have, but a little organization will go a long way in the future I think. (e.g. have a separate "it" or "describe" block for testing Pod service discovery, one for Service, and so on).

Reply:
There are also a lot of opportunities for using data tables here - see https://godoc.org/github.com/onsi/ginkgo/extensions/table. That will reduce the duplication - we can create a list of prometheus queries and their expected outcome for example.

Reply:
Should be better now!